### ISSUE TITLE
ModuleNotFoundError: No module named 'uaclient.entitlements'

### ISSUE THREAD

**Issue by zkytony** [canonical/ubuntu-advantage-client#1663]
I am getting the error "
```
ModuleNotFoundError: No module named 'uaclient.entitlements'
```
whenever I run apt upgrade or apt install commands. For example,
```
$ sudo apt upgrade
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] y
Setting up ubuntu-advantage-tools (27.0.2~18.04.1) ...
Traceback (most recent call last):
  File "<string>", line 2, in <module>
ModuleNotFoundError: No module named 'uaclient.entitlements'
dpkg: error processing package ubuntu-advantage-tools (--configure):
 installed ubuntu-advantage-tools package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 ubuntu-advantage-tools
E: Sub-process /usr/bin/dpkg returned an error code (1)
```
Everything was fine yesterday until I installed an Ubuntu base upgrade using the GUI software updater.

Before encountering the issue about "uaclient.entitlements", I got a ModuleNotFoundError for the 'uaclient` module. I resolved that by doing
```
pip3 install opcua-client
```
Now I have to resolve this one about "uaclient.entitlements". What should I do?

**Comment by blackboxsw** [comment]
@zkytony thank you for filing this bug. Firstly `opcua-client` module is a completely unrelated project for what looks like UA TCP communication if I'm not mistaken. So, pip3 installing that module only by happenstance shares the same module name :( as this project.  This is definitely not what you want to solve this problem.

It seems the python3 in your environment may not be from python3-minumal which by default has visiblility to a PYTHONPATH  which includes /usr/lib/python3/dist-packages/?  This directory is where ubuntu-advantage-tools deb package installs its python modules on disk. Is there a custom python3 setup in your environment?

**Comment by zkytony** [comment]
I see, thanks for the info! Wouldn't have known that `opcua-client` is unrelated.

Indeed, my python3's packages seem to go `'/usr/local/lib/python3.8/site-packages` when I print `sys.path`. And I do see `uaclient` under `/usr/lib/python3/dist-packages/`. I added `/usr/lib/python3/dist-packages/` to the PYTHONPATH by
```
export PYTHONPATH="${PYTHONPATH}:/usr/lib/pythno3/dist-packages/"
```
And I verified that `/usr/lib/pythno3/dist-packages` is in the list of paths in `sys.path`. However, I am still getting the error that 'uaclient' is not found

```
$ sudo apt-get upgrade
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] y
Setting up ubuntu-advantage-tools (27.0.2~18.04.1) ...
Traceback (most recent call last):
  File "<string>", line 2, in <module>
ModuleNotFoundError: No module named 'uaclient'
dpkg: error processing package ubuntu-advantage-tools (--configure):
 installed ubuntu-advantage-tools package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 ubuntu-advantage-tools
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by lucasmoura** [comment]
Hi @zkytony just to double check here, looking at the `PYTHONPATH` command, I see a typo on `/usr/lib/pythno3/dist-packages/` part, it should be `/usr/lib/python3/dist-packages/`. If that is the case, maybe fixing the typo will solve this issue.

**Comment by zkytony** [comment]
That indeed fixed it. Thanks for spotting the typo!!

**Comment by zkytony** [comment]
Thanks for spotting the typo! For a moment I thought it fixed the issue. But it actually didn't:
```
$ echo $PYTHONPATH
:/usr/lib/python3/dist-packages/
$ sudo apt-get upgrade
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] y
Setting up ubuntu-advantage-tools (27.0.2~18.04.1) ...
Traceback (most recent call last):
  File "<string>", line 2, in <module>
ModuleNotFoundError: No module named 'uaclient'
dpkg: error processing package ubuntu-advantage-tools (--configure):
 installed ubuntu-advantage-tools package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 ubuntu-advantage-tools
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by lucasmoura** [comment]
No problem @zkytony, I have a couple of ideas here:

* Do you have that path on your system ? `/usr/lib/python3/dist-packages` ? Maybe it should be `/usr/lib/python3.8/dist-packages/`
* I think that just adding `/usr/lib/python3.8` should be enough here.

Let me know if neither of those approaches work

**Comment by sisrfeng** [comment]
> Firstly opcua-client module is a completely unrelated project
Don't be confused by this unrelated one:
![image](https://user-images.githubusercontent.com/53520949/120401508-a95c0400-c372-11eb-8c76-cfd0b8f1a189.png)

this works:
export PYTHONPATH=/usr/lib/python3/dist-packages

this does not work:
/usr/lib/python3.6/site-packages

**Comment by blackboxsw** [comment]
This is also related to the following Launchpad bugs related to custom python3 installs not having a PYTHONPATH configured with visibility to  `/usr/lib/python3/dist-packages/`
 - https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1934077
 - https://bugs.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/+bug/1927348

**Comment by orndorffgrant** [comment]
This should be fixed as of 27.4.1 If you are experiencing this issue, please upgrade your package. If you still experience this issue after upgrading please re-open this issue.
